client/src/main/scala/org/http4s/client/Client.scala

```diff
@@ -1,84 +1,156 @@
-package org.http4s.client
+package org.http4s
+package client
 
-import org.http4s._
 import org.http4s.headers.Accept
 
 import scalaz.concurrent.Task
-
-trait Client {
-
-  /** Shutdown this client, closing any open connections and freeing resources */
-  def shutdown(): Task[Unit]
-
-  /** Prepare a single request
-    * @param req [[Request]] containing the headers, URI, etc.
-    * @return Task which will generate the Response
+import scalaz.stream.Process.eval_
+
+/**
+  * Contains a [[Response]] that needs to be disposed of to free the underlying
+  * HTTP connection.
+  * @param response
+  * @param dispose
+  */
+final case class DisposableResponse(response: Response, dispose: Task[Unit]) {
+  /**
+    * Returns a task to handle the response, safely disposing of the underlying
+    * HTTP connection when the task finishes.
     */
-  def prepare(req: Request): Task[Response]
+  def apply[A](f: Response => Task[A]): Task[A] =
+    f(response).onFinish { case _ => dispose }
+}
 
-  /** Prepare a single request
-    * @param req [[Request]] containing the headers, URI, etc.
-    * @return Task which will generate the Response
+/**
+  * A [[Client]] submits [[Request]]s to a server and processes the [[Response]].
+  *
+  * @param open a service to asynchronously return a [[DisposableResponse]] from
+  *             a [[Request]].  This is a low-level operation intended for client
+  *             implementations and middlewares.
+  *
+  * @param shutdown a Task to shut down this Shutdown this client, closing any
+  *                 open connections and freeing resources
+  */
+final case class Client(open: Service[Request, DisposableResponse], shutdown: Task[Unit]) {
+  /** Submits a request, and provides a callback to process the response.
+    *
+    * @param req The request to submit
+    * @param f A callback for the response to req.  The underlying HTTP connection
+    *          is disposed when the returned task completes.  Attempts to read the
+    *          response body afterward will result in an error.
+    * @return The result of applying f to the response to req
     */
-  final def apply(req: Request): Task[Response] = prepare(req)
+  def fetch[A](req: Request)(f: Response => Task[A]): Task[A] =
```

Comment:
This is perhaps the most common operation, and I'm tempted to call it `apply`.  Since this is a sweeping change and our client is popular, I did more than usual to provide a deprecation cycle for this 0.x project.

If we want to reclaim the name `apply`, we can discuss whether that's for 0.12 or 0.13.